Обзор методов показал, что одним из лучших методов обнаружения кругов является метод HoughCircles (https://docs.opencv.org/2.4/modules/imgproc/doc/feature_detection.html?highlight=houghcircles#houghcircles).

Он хорошо разобран в данной статье: https://www.pyimagesearch.com/2014/07/21/detecting-circles-images-using-opencv-hough-circles/

Применим его к нашей задаче

In [1]:
import cv2
import numpy as np

In [2]:
img = cv2.imread('task.jpg',0)
img = cv2.medianBlur(img,5)
cimg = cv2.cvtColor(img,cv2.COLOR_GRAY2BGR)

circles = cv2.HoughCircles(img,cv2.HOUGH_GRADIENT, dp = 1, minDist = 5,
                            param1=0.1,param2=0.1,minRadius=0,maxRadius=1)

circles = np.uint16(np.around(circles))
for i in circles[0,:]:
    cv2.circle(cimg,(i[0],i[1]),i[2],(0,255,0),2)
    #cv2.circle(cimg,(i[0],i[1]),2,(0,0,255),3)


cv2.imshow('detected circles',cimg)
cv2.waitKey(0)
cv2.destroyAllWindows()

Был проведен подбор параметров minDist и maxRadius с заранее подобранными и фиксированными dp,  param1, param2, minRadius.

После визуального подбора остановились на minDist = 5 и maxRadius=1. При значении maxRadius=2 точки также точно воспроизводятся.

In [3]:
circles.shape

(1L, 1084L, 3L)

Количество точек при maxRadius=1 и maxRadius=2 равны 1084 и 1107 соответственно. Оценим количество точек как 1100. 

**Ответ:** на картинке находится около 1100 точек.